File: tests/Unit/Boundary/TestReflectingBoundary.cc
#include "NodeList/SphNodeList.hh"
#include "Boundary/ReflectingBoundary.hh"
#include "Material/CGSUnits.hh"
#include "Material/PhysicalConstants.hh"
#include "Material/GammaLawGas.hh"

#include "Geometry/GeomPlane.cc"
#include "Geometry/GeomVector.cc"
#include "Geometry/GeomTensor.cc"
#include "NodeList/NodeList.cc"
#include "NodeList/FluidNodeList.cc"
#include "NodeList/SphNodeList.cc"
#include "Material/CGSUnits.cc"
#include "Material/MKSUnits.cc"
#include "Material/CosmologicalUnits.cc"
#include "Material/PhysicalConstants.cc"
#include "Material/GammaLawGas.cc"
#include "Neighbor/Neighbor.cc"
#include "Neighbor/NestedGridNeighbor.cc"
#include "Neighbor/GridCellIndex.cc"
#include "Neighbor/GridCellPlane.cc"
#include "Field/FieldBase.cc"
#include "Field/Field.cc"
#include "Boundary/Boundary.cc"
#include "Boundary/PlanarBoundary.cc"
#include "Boundary/ReflectingBoundary.cc"
#include "Kernel/Kernel.cc"
#include "Kernel/BSplineKernel.cc"
#include "Kernel/GaussianKernel.cc"
#include "Kernel/PiGaussianKernel.cc"
#include "Kernel/SuperGaussianKernel.cc"
#include "Kernel/W4SplineKernel.cc"

#include "Geometry/Dimension.cc"

int main() {

  int nxNodes = 10;
  int nyNodes = 10;
  int nzNodes = 10;
  int nNodes = nxNodes*nyNodes*nzNodes;

  double h = 1.0;
  double gamma = 5.0/3.0;
  double mu = 1.0;

  int neighborSearchType = 3; // GatherScatter
  int numGridLevels = 10;
  double topGridCellSize = 2.0;
  Dim<3>::Vector origin(0.0, 0.0, 0.0);
  double kernelExtent = 2.0;

  GammaLawGas<Dim<3>, PhysicalConstants<CGSUnits> > eos(gamma, mu);
  SphNodeList< Dim<3> > nodes(nNodes, eos);

  double dx = 1.0;
  double dy = 1.0;
  double dz = 1.0;

  int nxyNodes = nxNodes*nyNodes;
  Field<Dim<3>, Dim<3>::Vector> &positions = nodes.positions();
  Field<Dim<3>, Dim<3>::Scalar> &Hfield = nodes.Hfield();
  for (int i = 0; i < nzNodes; ++i) {
    for (int j = 0; j < nyNodes; ++j) {
      for (int k = 0; k < nxNodes; ++k) {
        int nodeID = i*nxyNodes + j*nxNodes + k;
        positions(nodeID) = Vector3d((k + 0.5)*dx,
                                     (j + 0.5)*dy,
                                     (i + 0.5)*dz);
        Hfield(nodeID) = h;
      }
    }
  }

  NestedGridNeighbor<Dim<3>, Dim<3>::Scalar>
    nestedNeighbor(nodes,
                   neighborSearchType,
                   numGridLevels,
                   topGridCellSize,
                   origin,
                   kernelExtent);
    

  GeomPlane<Dim<3> > xPlane0(Vector3d(0,0,0),
                                   Vector3d(1,0,0));
  GeomPlane<Dim<3> > xPlane1(Vector3d(10,0,0),
                                   Vector3d(-1,0,0));
  GeomPlane<Dim<3> > yPlane0(Vector3d(0,0,0),
                                   Vector3d(0,1,0));
  GeomPlane<Dim<3> > yPlane1(Vector3d(0,10,0),
                                   Vector3d(0,-1,0));
  GeomPlane<Dim<3> > zPlane0(Vector3d(0,0,0),
                                   Vector3d(0,0,1));
  GeomPlane<Dim<3> > zPlane1(Vector3d(0,0,10),
                                   Vector3d(0,0,-1));

  ReflectingBoundary<Dim<3>, SphNodeList<Dim<3> > > xbc0(xPlane0);
  ReflectingBoundary<Dim<3>, SphNodeList<Dim<3> > > xbc1(xPlane1);
  ReflectingBoundary<Dim<3>, SphNodeList<Dim<3> > > ybc0(yPlane0);
  ReflectingBoundary<Dim<3>, SphNodeList<Dim<3> > > ybc1(yPlane1);
  ReflectingBoundary<Dim<3>, SphNodeList<Dim<3> > > zbc0(zPlane0);
  ReflectingBoundary<Dim<3>, SphNodeList<Dim<3> > > zbc1(zPlane1);

  cerr << "nodes.numNodes, nodes.numInternalNodes, nodes.numGhostNodes: "
       << nodes.numNodes() << " "
       << nodes.numInternalNodes() << " "
       << nodes.numGhostNodes() << endl;

  cerr << "Applying x0 BC." << endl;
  xbc0.setGhostNodes(nodes);
  cerr << "nodes.numNodes, nodes.numInternalNodes, nodes.numGhostNodes: "
       << nodes.numNodes() << " "
       << nodes.numInternalNodes() << " "
       << nodes.numGhostNodes() << endl;
  
  cerr << "Applying x1 BC." << endl;
  xbc1.setGhostNodes(nodes);
  cerr << "nodes.numNodes, nodes.numInternalNodes, nodes.numGhostNodes: "
       << nodes.numNodes() << " "
       << nodes.numInternalNodes() << " "
       << nodes.numGhostNodes() << endl;
  
  cerr << "Applying y0 BC." << endl;
  ybc0.setGhostNodes(nodes);
  cerr << "nodes.numNodes, nodes.numInternalNodes, nodes.numGhostNodes: "
       << nodes.numNodes() << " "
       << nodes.numInternalNodes() << " "
       << nodes.numGhostNodes() << endl;
  
  cerr << "Applying y1 BC." << endl;
  ybc1.setGhostNodes(nodes);
  cerr << "nodes.numNodes, nodes.numInternalNodes, nodes.numGhostNodes: "
       << nodes.numNodes() << " "
       << nodes.numInternalNodes() << " "
       << nodes.numGhostNodes() << endl;
  
  cerr << "Applying z0 BC." << endl;
  zbc0.setGhostNodes(nodes);
  cerr << "nodes.numNodes, nodes.numInternalNodes, nodes.numGhostNodes: "
       << nodes.numNodes() << " "
       << nodes.numInternalNodes() << " "
       << nodes.numGhostNodes() << endl;
  
  cerr << "Applying z1 BC." << endl;
  zbc1.setGhostNodes(nodes);
  cerr << "nodes.numNodes, nodes.numInternalNodes, nodes.numGhostNodes: "
       << nodes.numNodes() << " "
       << nodes.numInternalNodes() << " "
       << nodes.numGhostNodes() << endl;
  
  return 1;
}

File: tests/Unit/Boundary/TestReflectingBoundary.py
from Spheral import *
from SpheralTestUtilities import *
from NeighborTestUtilities import *

################################################################################
title('3-D Sph node list')
nxNodes = 10
nyNodes = 10
nzNodes = 10
nNodes = nxNodes*nyNodes*nzNodes
h = 1.0/1.01

gamma = 5.0/3.0
mu = 1.0

neighborSearchType = 3 # GatherScatter
numGridLevels = 10
topGridCellSize = 2.5
origin = Vector3d(0.0, 0.0, 0.0)
kernelExtent = 2.0

eos = GammaLawGasCGS3d(gamma, mu)
nodes = SphNodeList3d(eos, nNodes)

dx = 1.0
dy = 1.0
dz = 1.0
nxyNodes = nxNodes*nyNodes
for i in xrange(nzNodes):
    for j in xrange(nyNodes):
        for k in xrange(nxNodes):
            iNode = i*nxyNodes + j*nxNodes + k
            nodes.positions[iNode] = Vector3d((k + 0.5)*dx,
                                              (j + 0.5)*dy,
                                              (i + 0.5)*dz)
            nodes.Hfield[iNode].xx = h
            nodes.Hfield[iNode].yy = h
            nodes.Hfield[iNode].zz = h

nestedNeighbor = NestedGridNeighbor3d(nodes,
                                      neighborSearchType,
                                      numGridLevels,
                                      topGridCellSize,
                                      origin,
                                      kernelExtent)

# Create reflecting boundary conditions.
xPlane0, xPlane1 = Plane3d((0,0,0), (1,0,0)), Plane3d((10,0,0), (-1,0,0))
yPlane0, yPlane1 = Plane3d((0,0,0), (0,1,0)), Plane3d((0,10,0), (0,-1,0))
zPlane0, zPlane1 = Plane3d((0,0,0), (0,0,1)), Plane3d((0,0,10), (0,0,-1))
xbc0, xbc1 = ReflectingBoundary3d(xPlane0), ReflectingBoundary3d(xPlane1)
ybc0, ybc1 = ReflectingBoundary3d(yPlane0), ReflectingBoundary3d(yPlane1)
zbc0, zbc1 = ReflectingBoundary3d(zPlane0), ReflectingBoundary3d(zPlane1)

output('xbc0, xbc1')
output('ybc0, ybc1')
output('zbc0, zbc1')

# Apply the boundary conditions to the node list.
print 'Applying boundary conditions to node list.'
output('nodes.numNodes, nodes.numInternalNodes, nodes.numGhostNodes')
output('xbc0.setNodeListGhostNodes(nodes)')
output('nodes.numNodes, nodes.numInternalNodes, nodes.numGhostNodes')
output('xbc1.setNodeListGhostNodes(nodes)')
output('nodes.numNodes, nodes.numInternalNodes, nodes.numGhostNodes')
output('ybc0.setNodeListGhostNodes(nodes)')
output('nodes.numNodes, nodes.numInternalNodes, nodes.numGhostNodes')
output('ybc1.setNodeListGhostNodes(nodes)')
output('nodes.numNodes, nodes.numInternalNodes, nodes.numGhostNodes')
output('zbc0.setNodeListGhostNodes(nodes)')
output('nodes.numNodes, nodes.numInternalNodes, nodes.numGhostNodes')
output('zbc1.setNodeListGhostNodes(nodes)')
output('nodes.numNodes, nodes.numInternalNodes, nodes.numGhostNodes')

neighborTimer = SpheralTimer("3-D nested neighbor timer")
neighborTimer.start()
output('nestedNeighbor.setMasterList(0)')
neighborTimer.stop()
output('nestedNeighbor.masterList[:]')
output('nestedNeighbor.coarseNeighborList[:]')
output('nestedNeighbor.setRefineNeighborList(0)')
output('nestedNeighbor.refineNeighborList[:]')

# Find the correct answer, and make sure all the actual neighbor nodes are
# included in the refined list.
checkTimer = SpheralTimer("3-D check neighbor timer")
checkTimer.start()
test = checkNeighbors(nestedNeighbor.refineNeighborList,
                      findNeighborNodes(nodes.positions[0],
                                        kernelExtent/nodes.Hfield[0].xx,
                                        nodes))
checkTimer.stop()

if test:
    print '3-D Sph Nested Neighbor test PASSED'
else:
    print '3-D Sph Nested Neighbor test FAILED'


neighborTimer.printStatus()
checkTimer.printStatus()

################################################################################
title('2-D Sph node list')
nxNodes = 10
nyNodes = 10
nNodes = nxNodes*nyNodes
h = 1.0/1.01

gamma = 5.0/3.0
mu = 1.0

neighborSearchType = 3 # GatherScatter
numGridLevels = 10
topGridCellSize = 2.5
origin = Vector2d(0.0, 0.0)
kernelExtent = 2.0

eos = GammaLawGasCGS2d(gamma, mu)
nodes = SphNodeList2d(eos, nNodes)

dx = 1.0
dy = 1.0
for i in xrange(nyNodes):
    for j in xrange(nxNodes):
        iNode = i*nxNodes + j
        nodes.positions[iNode] = Vector2d((j + 0.5)*dx,
                                          (i + 0.5)*dy)
        nodes.Hfield[iNode].xx = h
        nodes.Hfield[iNode].yy = h

nestedNeighbor = NestedGridNeighbor2d(nodes,
                                      neighborSearchType,
                                      numGridLevels,
                                      topGridCellSize,
                                      origin,
                                      kernelExtent)

# Create reflecting boundary conditions.
xPlane0, xPlane1 = Plane2d((0,0), (1,0)), Plane2d((10,0), (-1,0))
yPlane0, yPlane1 = Plane2d((0,0), (0,1)), Plane2d((0,10), (0,-1))
xbc0, xbc1 = ReflectingBoundary2d(xPlane0), ReflectingBoundary2d(xPlane1)
ybc0, ybc1 = ReflectingBoundary2d(yPlane0), ReflectingBoundary2d(yPlane1)

output('xbc0, xbc1')
output('ybc0, ybc1')

# Apply the boundary conditions to the node list.
print 'Applying boundary conditions to node list.'
output('nodes.numNodes, nodes.numInternalNodes, nodes.numGhostNodes')
output('xbc0.setNodeListGhostNodes(nodes)')
output('nodes.numNodes, nodes.numInternalNodes, nodes.numGhostNodes')
plotNodes(nodes, 0, 'After applying xbc0')
output('xbc1.setNodeListGhostNodes(nodes)')
output('nodes.numNodes, nodes.numInternalNodes, nodes.numGhostNodes')
plotNodes(nodes, 1, 'After applying xbc1')
output('ybc0.setNodeListGhostNodes(nodes)')
output('nodes.numNodes, nodes.numInternalNodes, nodes.numGhostNodes')
plotNodes(nodes, 2, 'After applying ybc0')
output('ybc1.setNodeListGhostNodes(nodes)')
output('nodes.numNodes, nodes.numInternalNodes, nodes.numGhostNodes')
plotNodes(nodes, 3, 'After applying ybc1')

neighborTimer = SpheralTimer("2-D nested neighbor timer")
neighborTimer.start()
output('nestedNeighbor.setMasterList(0)')
neighborTimer.stop()
output('nestedNeighbor.masterList[:]')
output('nestedNeighbor.coarseNeighborList[:]')
output('nestedNeighbor.setRefineNeighborList(0)')
output('nestedNeighbor.refineNeighborList[:]')

# Find the correct answer, and make sure all the actual neighbor nodes are
# included in the refined list.
checkTimer = SpheralTimer("2-D check neighbor timer")
checkTimer.start()
test = checkNeighbors(nestedNeighbor.refineNeighborList,
                      findNeighborNodes(nodes.positions[0],
                                        kernelExtent/nodes.Hfield[0].xx,
                                        nodes))
checkTimer.stop()

if test:
    print '2-D Sph Nested Neighbor test PASSED'
else:
    print '2-D Sph Nested Neighbor test FAILED'

neighborTimer.printStatus()
checkTimer.printStatus()

